### ISSUE TITLE
Explore removing deployment from the orchestration pipeline

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-jenkins-shared-library#439]
The orchestration pipeline works with phases: init, build, deploy, test, release, finalise.

When the target environment is DEV, the `build` phase executes the Jenkinsfile of each component. This actually builds AND deploys the component. When the target environment is QA/PROD, the Jenkinsfile of each component is not executed at all. The deployment of the templates/images is implemented in the orchestration pipeline phase `deploy`. That phase has a lot of duplication with the "deploy" part in the component Jenkinsfile.

Triggered by @metmajer suggestion to split build / deploy steps in order to be able to know if all components can build before we start deploying one of them, I started to wonder if it is possible to remove the "deploy" logic from the orchestration pipeline, and instead use the "deploy" logic within the component pipeline.

The basic idea goes something like this:

Component pipelines have two main blocks now:
```
odsComponentImportOpenShiftImageOrElse(...) { ... }
odsComponentRolloutOpenShiftDeployment(...)
```

What if, when the component pipeline is executed from the orchestration pipeline, we also pass down the orchestration phase in which we are at the moment, for example `build` or `deploy`. Now if we are in the `build` phase, we only execute `odsComponentImportOpenShiftImageOrElse` and skip `odsComponentRolloutOpenShiftDeployment`. And vice versa, if we are in the `deploy` phase, we only execute `odsComponentRolloutOpenShiftDeployment` and skip `odsComponentImportOpenShiftImageOrElse`.

Right now, when the target env is QA/PROD, the `deploy` phase of the orchestration pipeline also imports images, but we could refactor this. We could execute the `Jenkinsfile` also in the `build` phase when the target env is QA/PROD, and ensure we only import the image, and not allow to build it.

In effect, this change would shift quite some logic from the orchestration pipeline to the component pipeline. It would remove some duplication of code. My doubts however are that it could be brittle / surprising if the Jenkinsfile contains stages which are non-standard, and therefore do not care about the orchestration pipeline context. Not sure how we could deal with that yet. However, this might already be an issue as e.g. deployment is really standardised in the orchestration - if people do crazy stuff in the component pipeline, the orchestration deployment will like fail.

FYI @martsec @metmajer @clemensutschig

**Comment by michaelsauter** [comment]
> My doubts however are that it could be brittle / surprising if the Jenkinsfile contains stages which are non-standard, and therefore do not care about the orchestration pipeline context. Not sure how we could deal with that yet.

Maybe the better solution is to simply move the duplicated code into one service that is called from both places as opposed to executing the `Jenkinsfile` and running only parts from it.

**Comment by clemensutschig** [comment]
@michaelsauter  - I like the idea - but the complexity of refactoring (an already working pipeline) - which may have introduced a dependency and backwards compatibility make this very hard .. I vouch for the service, to use a much common ground as we can - and make the uptake of version as painless as possible .. m2c only though

**Comment by michaelsauter** [comment]
Yes, service sounds good, and I also would like to keep refactoring at a minimum. Low prio I think ...

**Comment by michaelsauter** [comment]
Newer thoughts on this are in https://github.com/opendevstack/ods-jenkins-shared-library/issues/479. I think overall, we do not actually want to give control over the deployment to custom code. For the orchestration pipeline, we do want to have control over the process so that everything happens according to the validated process. Therefore, I'm closing this in favour of #479.

To summarise:
* Component Pipeline = full flexibility
* Orchestration Pipeline = validated process
